README.md
# openevse-esp8266-emoncms
OpenEVSE posting to emoncms via ESP8266


Prerequisites
===========

esp-open-sdk: https://github.com/pfalcon/esp-open-sdk

## Debian wheezy

Install tools
```
$ sudo apt-get install make unrar-free autoconf automake libtool gcc g++ gperf flex bison texinfo gawk ncurses-dev libexpat-dev python python-serial sed git libtool-bin
```

Install esp-open-sdk

```
$ git clone --recursive https://github.com/pfalcon/esp-open-sdk.git
$ unset LD_LIBRARY_PATH
$ make STANDALONE=n
```
Go get yourself a big cup of coffee... esp-open-sdk takes a looonnngg time to build.

Set path to esp-open-sdk and esptool

```
$ export PATH=<path-to-esp-open-sdk>/xtensa-lx106-elf/bin:<path-to-esp-open-sdk>/esptool:$PATH
```

Building openevse-esp8266-emoncms
=================================

1. edit Makefile and set ESP_OPEN_SDK to point to path to your esp-open-sdk

2. create user/user_config.h and add:
```
#define SSID "YOUR SSID"
#define PASS "YOUR PASSPHRASE"
#define NODE "0"
#define PRIVATEKEY "YOUR EMONCMS APIKEY"
```

3. Build firmware

```
 $ make
```

4. Flash into ESP8266
```
 $ sudo make flash
```









CHANGELOG

v0.0.4 SCL 20151026
- added isIntegerString() - don't save any parameters unless they are integer
- call rapiProcessReply(0) before sending commands to clear out rx buffer
	
v0.0.3 SCL 20150926
- try again to fix corrupted data after several hours
  -> split out rapiProcessReply()
  -> extern rapiTokenCnt in rapi.h
  -> minor tweaks

v0.0.2 SCL 20150924
- volts & amps getting corrupted -> hardcoded to 16 after running several hours.
  no idea why. Change variables to strings to get rid of itoa()/ltoa()

v0.0.1 SCL 20150924
- initial checkin


user/rapi.c
/*
 * Copyright (c) 2015 Sam C. Lin
 *
 * This file is part of Open EVSE.
 * Open EVSE is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3, or (at your option)
 * any later version.
 * Open EVSE is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License
 * along with Open EVSE; see the file COPYING.  If not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

#include "ets_sys.h"
#include "driver/uart.h"
#include "osapi.h"
#include "os_type.h"
#include "user_config.h"
#include "rapi.h"


char respBuf[RAPI_BUFLEN];
int rapiTokenCnt;
char *rapiTokens[RAPI_MAX_TOKENS];

char ICACHE_FLASH_ATTR *utoa(unsigned num, char *str, int radix) {
    char temp[17];  //an int can only be 16 bits long
                    //at radix 2 (binary) the string
                    //is at most 16 + 1 null long.
    int temp_loc = 0;
    int digit;
    int str_loc = 0;

    //construct a backward string of the number.
    do {
        digit = (unsigned int)num % radix;
        if (digit < 10) 
            temp[temp_loc++] = digit + '0';
        else
            temp[temp_loc++] = digit - 10 + 'A';
        num = ((unsigned int)num) / radix;
    } while ((unsigned int)num > 0);

    temp_loc--;


    //now reverse the string.
    while ( temp_loc >=0 ) {// while there are still chars
        str[str_loc++] = temp[temp_loc--];    
    }
    str[str_loc] = 0; // add null termination.

    return str;
}

void ICACHE_FLASH_ATTR rapiInit()
{
  *respBuf = 0;
}


void ICACHE_FLASH_ATTR rapiSendCmd(const char *cmdstr)
{
  rapiProcessReply(0); // clear out rx buffer
  uart0_sendStr(cmdstr);
  const char *s = cmdstr;
  uint8_t chk = 0;
  while (*s) {
    chk ^= *(s++);
  }
  respBuf[0] = '^';
  utoa(chk,respBuf+1,16);
  strcat(respBuf,"\r");
  uart0_sendStr(respBuf);

  *respBuf = 0;
  rapiTokenCnt = 0;
}


void ICACHE_FLASH_ATTR tokenize()
{
  char *s = respBuf;
  rapiTokenCnt = 0;
  while (*s) {
    rapiTokens[rapiTokenCnt++] = s++;
    if (rapiTokenCnt == RAPI_MAX_TOKENS) break;
    while (*s && (*s != ' ')) s++;
    if (*s == ' ') *(s++) = '\0';
  }
}

/*
 * return values:
 * -1= timeout
 * 0= success
 * 1=$NK
 * 2=invalid RAPI response
*/
int ICACHE_FLASH_ATTR rapiProcessReply(uint32 timeout)
{
 start:
  rapiTokenCnt = 0;
  *respBuf = 0;

  int bufpos = 0;
  uint32 mss = system_get_time();
  do {
    int rxc = uart0_rx_one_char();
    if (rxc == -1) {
      os_delay_us(10);
    }
    else {
      char c = rxc;

      if (!bufpos && (c != '$')) {
	// wait for start character
	continue;
      }
      else if (c == '\r') {
	respBuf[bufpos] = '\0';
	tokenize();
      }
      else {
	if (c == '$') bufpos = 0;
	respBuf[bufpos++] = c;
	// buffer full and no termination 
	if (bufpos >= (RAPI_BUFLEN-1)) goto start;
      }
    }
  } while (!rapiTokenCnt && ((system_get_time() - mss) < timeout));
  
/*  
    dbgprint("\n\rTOKENCNT: ");dbgprintln(rapiTokenCnt);
    for (int i=0;i < rapiTokenCnt;i++) {
      dbgprintln(rapiTokens[i]);
    }
    dbgprintln("");
*/
  

  if (rapiTokenCnt) {
    if (!strcmp(respBuf,"$OK")) {
      return 0;
    }
    else if (!strcmp(respBuf,"$NK")) {
      return 1;
    }
    else if (!strcmp(respBuf,"$WF")) { // async WIFI
      goto start;
    }
    else if (!strcmp(respBuf,"$ST")) { // async EVSE state transition
      goto start;
    }
    else {
      return 2;
    }
  }
  else {
    return -1;
  }
}


user/rapi.h
/*
 * Copyright (c) 2015 Sam C. Lin
 *
 * This file is part of Open EVSE.
 * Open EVSE is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3, or (at your option)
 * any later version.
 * Open EVSE is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License
 * along with Open EVSE; see the file COPYING.  If not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

#pragma once

#define RAPI_TIMEOUT_US 250000ul
#define RAPI_BUFLEN 32
#define RAPI_MAX_TOKENS 5

extern char *rapiTokens[RAPI_MAX_TOKENS];
extern int rapiTokenCnt;

void ICACHE_FLASH_ATTR rapiInit();
void ICACHE_FLASH_ATTR rapiSendCmd(const char *cmdstr);
int ICACHE_FLASH_ATTR rapiProcessReply(uint32 timeout);


driver/uart.c
#if 0
/******************************************************************************
 * FunctionName : uart0_rx_intr_handler
 * Description  : Internal used function
 *                UART0 interrupt handler, add self handle code inside
 * Parameters   : void *para - point to ETS_UART_INTR_ATTACH's arg
 * Returns      : NONE
*******************************************************************************/
LOCAL void
uart0_rx_intr_handler(void *para)
{
    /* uart0 and uart1 intr combine togther, when interrupt occur, see reg 0x3ff20020, bit2, bit0 represents
     * uart1 and uart0 respectively
     */
    RcvMsgBuff *pRxBuff = (RcvMsgBuff *)para;
    uint8 RcvChar;

    if (UART_RXFIFO_FULL_INT_ST != (READ_PERI_REG(UART_INT_ST(UART0)) & UART_RXFIFO_FULL_INT_ST)) {
        return;
    }

    WRITE_PERI_REG(UART_INT_CLR(UART0), UART_RXFIFO_FULL_INT_CLR);

    while (READ_PERI_REG(UART_STATUS(UART0)) & (UART_RXFIFO_CNT << UART_RXFIFO_CNT_S)) {
        RcvChar = READ_PERI_REG(UART_FIFO(UART0)) & 0xFF;

        /* you can add your handle code below.*/

        *(pRxBuff->pWritePos) = RcvChar;

        // insert here for get one command line from uart
        if (RcvChar == '\r') {
            pRxBuff->BuffState = WRITE_OVER;
        }

        pRxBuff->pWritePos++;

        // if we hit the end of the buffer, loop back to the beginning
        if (pRxBuff->pWritePos == (pRxBuff->pRcvMsgBuff + RX_BUFF_SIZE)) {
            // overflow ...we may need more error handle here.
            pRxBuff->pWritePos = pRxBuff->pRcvMsgBuff ;
        }
    }
   system_os_post(user_procTaskPrio, 0, 0 );
}






#endif



/******************************************************************************
 * Copyright 2013-2014 Espressif Systems (Wuxi)
 *
 * FileName: uart.c
 *
 * Description: Two UART mode configration and interrupt handler.
 *              Check your hardware connection while use this mode.
 *
 * Modification history:
 *     2014/3/12, v1.0 create this file.
*******************************************************************************/
#include "ets_sys.h"
#include "osapi.h"
#include "driver/uart.h"
#include "osapi.h"
#include "driver/uart_register.h"
//#include "ssc.h"

#define user_procTaskPrio        1

// UartDev is defined and initialized in rom code.
extern UartDevice    UartDev;
//extern os_event_t    at_recvTaskQueue[at_recvTaskQueueLen];

LOCAL void uart0_rx_intr_handler(void *para);

int uart0_rx_one_char() {
  if(UartDev.rcv_buff.pReadPos == UartDev.rcv_buff.pWritePos) return -1;
  int ret = *UartDev.rcv_buff.pReadPos;
  UartDev.rcv_buff.pReadPos++;
  if(UartDev.rcv_buff.pReadPos == (UartDev.rcv_buff.pRcvMsgBuff + RX_BUFF_SIZE)) {
    UartDev.rcv_buff.pReadPos = UartDev.rcv_buff.pRcvMsgBuff;
  }

  return ret;
}


/******************************************************************************
 * FunctionName : uart1_tx_one_char
 * Description  : Internal used function
 *                Use uart1 interface to transfer one char
 * Parameters   : uint8 TxChar - character to tx
 * Returns      : OK
*******************************************************************************/
LOCAL STATUS
uart_tx_one_char(uint8 uart, uint8 TxChar)
{
    while (true)
    {
      uint32 fifo_cnt = READ_PERI_REG(UART_STATUS(uart)) & (UART_TXFIFO_CNT<<UART_TXFIFO_CNT_S);
      if ((fifo_cnt >> UART_TXFIFO_CNT_S & UART_TXFIFO_CNT) < 126) {
        break;
      }
    }

    WRITE_PERI_REG(UART_FIFO(uart) , TxChar);
    return OK;
}

/******************************************************************************
 * FunctionName : uart0_sendStr
 * Description  : use uart0 to transfer buffer
 * Parameters   : uint8 *buf - point to send buffer
 *                uint16 len - buffer len
 * Returns      :
*******************************************************************************/
void ICACHE_FLASH_ATTR
uart0_sendStr(const char *str)
{
	while(*str)
	{
		uart_tx_one_char(UART0, *str++);
	}
}

#if 0
/******************************************************************************
 * FunctionName : uart_config
 * Description  : Internal used function
 *                UART0 used for data TX/RX, RX buffer size is 0x100, interrupt enabled
 *                UART1 just used for debug output
 * Parameters   : uart_no, use UART0 or UART1 defined ahead
 * Returns      : NONE
*******************************************************************************/
LOCAL void ICACHE_FLASH_ATTR
uart_config(uint8 uart_no)
{
  if (uart_no == UART1)
  {
    PIN_FUNC_SELECT(PERIPHS_IO_MUX_GPIO2_U, FUNC_U1TXD_BK);
  }
  else
  {
    /* rcv_buff size if 0x100 */
    ETS_UART_INTR_ATTACH(uart0_rx_intr_handler,  &(UartDev.rcv_buff));
    PIN_PULLUP_DIS(PERIPHS_IO_MUX_U0TXD_U);
    PIN_FUNC_SELECT(PERIPHS_IO_MUX_U0TXD_U, FUNC_U0TXD);
    PIN_FUNC_SELECT(PERIPHS_IO_MUX_MTDO_U, FUNC_U0RTS);
  }

  uart_div_modify(uart_no, UART_CLK_FREQ / (UartDev.baut_rate));

  WRITE_PERI_REG(UART_CONF0(uart_no), UartDev.exist_parity
                 | UartDev.parity
                 | (UartDev.stop_bits << UART_STOP_BIT_NUM_S)
                 | (UartDev.data_bits << UART_BIT_NUM_S));

  //clear rx and tx fifo,not ready
  SET_PERI_REG_MASK(UART_CONF0(uart_no), UART_RXFIFO_RST | UART_TXFIFO_RST);
  CLEAR_PERI_REG_MASK(UART_CONF0(uart_no), UART_RXFIFO_RST | UART_TXFIFO_RST);

  //set rx fifo trigger
//  WRITE_PERI_REG(UART_CONF1(uart_no),
//                 ((UartDev.rcv_buff.TrigLvl & UART_RXFIFO_FULL_THRHD) << UART_RXFIFO_FULL_THRHD_S) |
//                 ((96 & UART_TXFIFO_EMPTY_THRHD) << UART_TXFIFO_EMPTY_THRHD_S) |
//                 UART_RX_FLOW_EN);
  if (uart_no == UART0)
  {
    //set rx fifo trigger
    WRITE_PERI_REG(UART_CONF1(uart_no),
                   ((0x10 & UART_RXFIFO_FULL_THRHD) << UART_RXFIFO_FULL_THRHD_S) |
                   ((0x10 & UART_RX_FLOW_THRHD) << UART_RX_FLOW_THRHD_S) |
                   UART_RX_FLOW_EN |
                   (0x02 & UART_RX_TOUT_THRHD) << UART_RX_TOUT_THRHD_S |
                   UART_RX_TOUT_EN);
    SET_PERI_REG_MASK(UART_INT_ENA(uart_no), UART_RXFIFO_TOUT_INT_ENA |
                      UART_FRM_ERR_INT_ENA);
  }
  else
  {
    WRITE_PERI_REG(UART_CONF1(uart_no),
                   ((UartDev.rcv_buff.TrigLvl & UART_RXFIFO_FULL_THRHD) << UART_RXFIFO_FULL_THRHD_S));
  }

  //clear all interrupt
  WRITE_PERI_REG(UART_INT_CLR(uart_no), 0xffff);
  //enable rx_interrupt
  SET_PERI_REG_MASK(UART_INT_ENA(uart_no), UART_RXFIFO_FULL_INT_ENA);
}
#endif

/******************************************************************************
 * FunctionName : uart_config
 * Description  : Internal used function
 *                UART0 used for data TX/RX, RX buffer size is 0x100, interrupt enabled
 *                UART1 just used for debug output
 * Parameters   : uart_no, use UART0 or UART1 defined ahead
 * Returns      : NONE
*******************************************************************************/
LOCAL void ICACHE_FLASH_ATTR
uart_config(uint8 uart_no)
{
    if (uart_no == UART1) {
        PIN_FUNC_SELECT(PERIPHS_IO_MUX_GPIO2_U, FUNC_U1TXD_BK);
    } else {
        /* rcv_buff size if 0x100 */
        ETS_UART_INTR_ATTACH(uart0_rx_intr_handler,  &(UartDev.rcv_buff));
        PIN_PULLUP_DIS(PERIPHS_IO_MUX_U0TXD_U);
        PIN_FUNC_SELECT(PERIPHS_IO_MUX_U0TXD_U, FUNC_U0TXD);
    }

    uart_div_modify(uart_no, UART_CLK_FREQ / (UartDev.baut_rate));

    WRITE_PERI_REG(UART_CONF0(uart_no),    UartDev.exist_parity
                   | UartDev.parity
                   | (UartDev.stop_bits << UART_STOP_BIT_NUM_S)
                   | (UartDev.data_bits << UART_BIT_NUM_S));


    //clear rx and tx fifo,not ready
    SET_PERI_REG_MASK(UART_CONF0(uart_no), UART_RXFIFO_RST | UART_TXFIFO_RST);
    CLEAR_PERI_REG_MASK(UART_CONF0(uart_no), UART_RXFIFO_RST | UART_TXFIFO_RST);

    //set rx fifo trigger
    WRITE_PERI_REG(UART_CONF1(uart_no), (UartDev.rcv_buff.TrigLvl & UART_RXFIFO_FULL_THRHD) << UART_RXFIFO_FULL_THRHD_S);

    //clear all interrupt
    WRITE_PERI_REG(UART_INT_CLR(uart_no), 0xffff);
    //enable rx_interrupt
    SET_PERI_REG_MASK(UART_INT_ENA(uart_no), UART_RXFIFO_FULL_INT_ENA);
}


/******************************************************************************
 * FunctionName : uart1_write_char
 * Description  : Internal used function
 *                Do some special deal while tx char is '\r' or '\n'
 * Parameters   : char c - character to tx
 * Returns      : NONE
*******************************************************************************/
void ICACHE_FLASH_ATTR
uart1_write_char(char c)
{
  if (c == '\n')
  {
    uart_tx_one_char(UART1, '\r');
    uart_tx_one_char(UART1, '\n');
  }
  else if (c == '\r')
  {
  }
  else
  {
    uart_tx_one_char(UART1, c);
  }
}



void ICACHE_FLASH_ATTR
uart0_write_char(char c)
{
  if (c == '\n')
  {
    uart_tx_one_char(UART0, '\r');
    uart_tx_one_char(UART0, '\n');
  }
  else if (c == '\r')
  {
  }
  else
  {
    uart_tx_one_char(UART0, c);
  }
}
/******************************************************************************
 * FunctionName : uart0_tx_buffer
 * Description  : use uart0 to transfer buffer
 * Parameters   : uint8 *buf - point to send buffer
 *                uint16 len - buffer len
 * Returns      :
*******************************************************************************/
void ICACHE_FLASH_ATTR
uart0_tx_buffer(uint8 *buf, uint16 len)
{
  uint16 i;

  for (i = 0; i < len; i++)
  {
    uart_tx_one_char(UART0, buf[i]);
  }
}



/******************************************************************************
 * FunctionName : uart0_rx_intr_handler
 * Description  : Internal used function
 *                UART0 interrupt handler, add self handle code inside
 * Parameters   : void *para - point to ETS_UART_INTR_ATTACH's arg
 * Returns      : NONE
*******************************************************************************/
//extern void at_recvTask(void);

LOCAL void
uart0_rx_intr_handler(void *para)
{

  /* uart0 and uart1 intr combine togther, when interrupt occur, see reg 0x3ff20020, bit2, bit0 represents
    * uart1 and uart0 respectively
    */
//  RcvMsgBuff *pRxBuff = (RcvMsgBuff *)para;
  uint8 RcvChar;
  uint8 uart_no = UART0;//UartDev.buff_uart_no;
#if 0
//  if (UART_RXFIFO_FULL_INT_ST != (READ_PERI_REG(UART_INT_ST(uart_no)) & UART_RXFIFO_FULL_INT_ST))
//  {
//    return;
//  }
//  if (UART_RXFIFO_FULL_INT_ST == (READ_PERI_REG(UART_INT_ST(uart_no)) & UART_RXFIFO_FULL_INT_ST))
//  {
////    at_recvTask();
//    RcvChar = READ_PERI_REG(UART_FIFO(uart_no)) & 0xFF;
//    system_os_post(at_recvTaskPrio, NULL, RcvChar);
//    WRITE_PERI_REG(UART_INT_CLR(uart_no), UART_RXFIFO_FULL_INT_CLR);
//  }
  if(UART_FRM_ERR_INT_ST == (READ_PERI_REG(UART_INT_ST(uart_no)) & UART_FRM_ERR_INT_ST))
  {
    os_printf("FRM_ERR\r\n");
    WRITE_PERI_REG(UART_INT_CLR(uart_no), UART_FRM_ERR_INT_CLR);
  }

  if(UART_RXFIFO_FULL_INT_ST == (READ_PERI_REG(UART_INT_ST(uart_no)) & UART_RXFIFO_FULL_INT_ST))
  {
//    os_printf("fifo full\r\n");
    ETS_UART_INTR_DISABLE();/////////

    //system_os_post(at_recvTaskPrio, 0, 0);

//    WRITE_PERI_REG(UART_INT_CLR(uart_no), UART_RXFIFO_FULL_INT_CLR);
//    while (READ_PERI_REG(UART_STATUS(uart_no)) & (UART_RXFIFO_CNT << UART_RXFIFO_CNT_S))
//    {
////      at_recvTask();
//      RcvChar = READ_PERI_REG(UART_FIFO(uart_no)) & 0xFF;
//      system_os_post(at_recvTaskPrio, NULL, RcvChar);
//    }
  }
  else if(UART_RXFIFO_TOUT_INT_ST == (READ_PERI_REG(UART_INT_ST(uart_no)) & UART_RXFIFO_TOUT_INT_ST))
  {
    ETS_UART_INTR_DISABLE();/////////

    //system_os_post(at_recvTaskPrio, 0, 0);

//    WRITE_PERI_REG(UART_INT_CLR(uart_no), UART_RXFIFO_TOUT_INT_CLR);
////    os_printf("rx time over\r\n");
//    while (READ_PERI_REG(UART_STATUS(uart_no)) & (UART_RXFIFO_CNT << UART_RXFIFO_CNT_S))
//    {
////      os_printf("process recv\r\n");
////      at_recvTask();
//      RcvChar = READ_PERI_REG(UART_FIFO(uart_no)) & 0xFF;
//      system_os_post(at_recvTaskPrio, NULL, RcvChar);
//    }
  }

//  WRITE_PERI_REG(UART_INT_CLR(uart_no), UART_RXFIFO_FULL_INT_CLR);

//  if (READ_PERI_REG(UART_STATUS(uart_no)) & (UART_RXFIFO_CNT << UART_RXFIFO_CNT_S))
//  {
//    RcvChar = READ_PERI_REG(UART_FIFO(uart_no)) & 0xFF;
//    at_recvTask();
//    *(pRxBuff->pWritePos) = RcvChar;

//    system_os_post(at_recvTaskPrio, NULL, RcvChar);

//    //insert here for get one command line from uart
//    if (RcvChar == '\r')
//    {
//      pRxBuff->BuffState = WRITE_OVER;
//    }
//
//    pRxBuff->pWritePos++;
//
//    if (pRxBuff->pWritePos == (pRxBuff->pRcvMsgBuff + RX_BUFF_SIZE))
//    {
//      // overflow ...we may need more error handle here.
//      pRxBuff->pWritePos = pRxBuff->pRcvMsgBuff ;
//    }
//  }

  /* uart0 and uart1 intr combine togther, when interrupt occur, see reg 0x3ff20020, bit2, bit0 represents
    * uart1 and uart0 respectively
    */
#endif
   RcvMsgBuff *pRxBuff = (RcvMsgBuff *)para;
#if 0
   uint8 RcvChar;
#endif

   if (UART_RXFIFO_FULL_INT_ST != (READ_PERI_REG(UART_INT_ST(UART0)) & UART_RXFIFO_FULL_INT_ST)) {
       return;
   }

   WRITE_PERI_REG(UART_INT_CLR(UART0), UART_RXFIFO_FULL_INT_CLR);

   while (READ_PERI_REG(UART_STATUS(UART0)) & (UART_RXFIFO_CNT << UART_RXFIFO_CNT_S)) {
       RcvChar = READ_PERI_REG(UART_FIFO(UART0)) & 0xFF;

       /* you can add your handle code below.*/

       *(pRxBuff->pWritePos) = RcvChar;

       // insert here for get one command line from uart
       if (RcvChar == '\r') {
           pRxBuff->BuffState = WRITE_OVER;
       }

       pRxBuff->pWritePos++;

       // if we hit the end of the buffer, loop back to the beginning
       if (pRxBuff->pWritePos == (pRxBuff->pRcvMsgBuff + RX_BUFF_SIZE)) {
           // overflow ...we may need more error handle here.
           pRxBuff->pWritePos = pRxBuff->pRcvMsgBuff ;
       }
   }
  system_os_post(user_procTaskPrio, 0, 0 );
//	system_os_post(MQTT_TASK_PRIO, 0, (os_param_t)mqttClient);

}

/******************************************************************************
 * FunctionName : uart_init
 * Description  : user interface for init uart
 * Parameters   : UartBautRate uart0_br - uart0 bautrate
 *                UartBautRate uart1_br - uart1 bautrate
 * Returns      : NONE
*******************************************************************************/
void ICACHE_FLASH_ATTR
uart_init(UartBautRate uart0_br, UartBautRate uart1_br)
{
  // rom use 74880 baut_rate, here reinitialize
  UartDev.baut_rate = uart0_br;
  uart_config(UART0);
  UartDev.baut_rate = uart1_br;
  uart_config(UART1);
  ETS_UART_INTR_ENABLE();

  // install uart1 putc callback
  os_install_putc1((void *)uart0_write_char);
}


void ICACHE_FLASH_ATTR
uart_reattach()
{
	uart_init(BIT_RATE_74880, BIT_RATE_74880);
//  ETS_UART_INTR_ATTACH(uart_rx_intr_handler_ssc,  &(UartDev.rcv_buff));
//  ETS_UART_INTR_ENABLE();
}


user/user_main.c
/*
 * Copyright (c) 2015 Sam C. Lin
 *
 * This file is part of Open EVSE.
 * Open EVSE is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3, or (at your option)
 * any later version.
 * Open EVSE is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License
 * along with Open EVSE; see the file COPYING.  If not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */


#include "ets_sys.h"
#include "driver/uart.h"
#include "osapi.h"
#include "os_type.h"
#include "user_config.h"
#include "user_interface.h"
#include "rapi.h"
#include "httpclient.h"

//#define dbghttp
//#define dbgfakedata

char ssid[33] = SSID;
char pass[65] = PASS;
char privateKey[33] = PRIVATEKEY;
char node[5] = NODE;
const char *url = "http://data.openevse.com/emoncms/input/post.json?";
const char* inputID_AMP   = "OpenEVSE_AMP:";
const char* inputID_VOLT   = "OpenEVSE_VOLT:";
const char* inputID_TEMP1   = "OpenEVSE_TEMP1:";
const char* inputID_TEMP2   = "OpenEVSE_TEMP2:";
const char* inputID_TEMP3   = "OpenEVSE_TEMP3:";
const char* inputID_PILOT   = "OpenEVSE_PILOT:";


#define user_procTaskPrio        0
#define user_procTaskQueueLen    1
os_event_t    user_procTaskQueue[user_procTaskQueueLen];
static void user_procTask(os_event_t *events);

static volatile os_timer_t some_timer;


char pilot[5];
char amp[10];
char volt[10];
char temp1[5];
char temp2[5];
char temp3[5];

void httpCallback(char * response, int http_status, char * full_response)
{
#ifdef dbghttp
    os_printf("http_status=%d\n", http_status);
    if (http_status != HTTP_STATUS_GENERIC_ERROR) {
        os_printf("strlen(full_response)=%d\n", strlen(full_response));
        os_printf("response=%s<EOF>\n", response);
    }
#endif
}

int tCnt;

int isIntegerString(const char *s)
{
  while (*s) {
    if ((*s != '-') && !isdigit(*s)) return 0;
    s++;
  }

  return 1;
}

void ICACHE_FLASH_ATTR some_timerfunc(void *arg)
{
  tCnt++;
  if ((wifi_station_get_connect_status() != STATION_GOT_IP) || (tCnt == 50)) {
    tCnt = -1;
  }
  else {
    if (tCnt == 0) {
      *volt = 0;
      os_strcpy(amp,"0");
      *pilot = 0;
      *temp1 = 0;
      *temp2 = 0;
      *temp3 = 0;
      rapiSendCmd("$GE");
    }
    else if (tCnt == 1) {
      rapiProcessReply(RAPI_TIMEOUT_US);
      if ((rapiTokenCnt == 3) && isIntegerString(rapiTokens[1])) {
	os_strcpy(pilot,rapiTokens[1]);
      }
    }
    else if (tCnt == 2) {
      rapiSendCmd("$GG");
    }
    else if (tCnt == 3) {
      rapiProcessReply(RAPI_TIMEOUT_US);
      if (rapiTokenCnt == 3) {
	if (isIntegerString(rapiTokens[1])) {
	    os_strcpy(amp,rapiTokens[1]);
	}
	if (isIntegerString(rapiTokens[2])) {
	  os_strcpy(volt,rapiTokens[2]);
	  if (*volt == '-') *volt = 0;
	}
      }
    }
    else if (tCnt == 4) {
      rapiSendCmd("$GP");
    }
    else if (tCnt == 5) {
      rapiProcessReply(RAPI_TIMEOUT_US);
      if (rapiTokenCnt == 4) {
	if (isIntegerString(rapiTokens[1])) {
	  os_strcpy(temp1,rapiTokens[1]);
	}
	if (isIntegerString(rapiTokens[2])) {
	  os_strcpy(temp2,rapiTokens[2]);
	}
	if (isIntegerString(rapiTokens[3])) {
	  os_strcpy(temp3,rapiTokens[3]);
	}
      }
    }
    else if (tCnt == 6) {
      char req[300];

#ifdef dbgfakedata
      os_strcpy(node,"99");
      os_strcpy(volt,"238945");
      os_strcpy(amp,"16023");
      os_strcpy(pilot,"16");
      os_strcpy(temp1,"136");
      os_strcpy(temp2,"235");
      os_strcpy(temp3,"335");
#endif

      os_strcpy(req,url);
      os_strcat(req,"node=");
      os_strcat(req,node);
      os_strcat(req,"&apikey=");
      os_strcat(req,privateKey);
      os_strcat(req,"&json={");
      
      os_strcat(req,inputID_AMP);
      os_strcat(req,amp);

      if (*pilot) {
	os_strcat(req,",");
	os_strcat(req,inputID_PILOT);
	os_strcat(req,pilot);
      }
      
      if (*volt) {
	os_strcat(req,",");
	os_strcat(req,inputID_VOLT);
	os_strcat(req,volt);
      }
      if (*temp1 > '0') {
	os_strcat(req,",");
	os_strcat(req,inputID_TEMP1);
	os_strcat(req,temp1);
      }
      if (*temp2 > '0') {
	os_strcat(req,",");
	os_strcat(req,inputID_TEMP2);
	os_strcat(req,temp2);
      }
      if (*temp3 >'0') {
	os_strcat(req,",");
	os_strcat(req,inputID_TEMP3);
	os_strcat(req,temp3);
      }
      os_strcat(req,"}");
      
#ifdef dbghttp
    os_printf("\n%s\n",req);
#endif
      http_get(req,"",httpCallback);
    }
  }
}

//Do nothing function
static void ICACHE_FLASH_ATTR
user_procTask(os_event_t *events)
{
  //    os_delay_us(10);
}

//Init function 
void ICACHE_FLASH_ATTR
user_init()
{
  tCnt = -1;

  struct station_config stationConf;
  
  //Set station mode
  wifi_set_opmode(STATION_MODE);

    
  //Set ap settings
  stationConf.bssid_set = 0; //need not check MAC address of AP
  os_memcpy(&stationConf.ssid, ssid, 32);
  os_memcpy(&stationConf.password, pass, 64);
  wifi_station_set_config(&stationConf);

  rapiInit();

  uart_init(BIT_RATE_115200, BIT_RATE_115200);

  //Disarm timer
  os_timer_disarm(&some_timer);
  
  //Setup timer
  os_timer_setfn(&some_timer, (os_timer_func_t *)some_timerfunc, NULL);
  
  //Arm the timer
  //&some_timer is the pointer
  //1000 is the fire time in ms
  //0 for once and 1 for repeating
  os_timer_arm(&some_timer, 500, 1);
  
  //Start os task
  system_os_task(user_procTask, user_procTaskPrio,user_procTaskQueue, user_procTaskQueueLen);
}
